ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -94,56 +98,60 @@ private[apiserver] final class ApiPackageManagementService private (
+  private def decodeAndValidate(
+      stream: ZipInputStream
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Try[Dar[Archive]] =
     for {
-      dar <- darReader.readArchive("package-upload", stream).toTry
-      packages <- dar.all.traverse(Decode.decodeArchive(_)).toTry
+      dar <- darReader
+        .readArchive("package-upload", stream)
+        .loggingValidation(Validation.handleLfArchiveError)
+      packages <- dar.all
+        .traverse(Decode.decodeArchive(_))
+        .loggingValidation(Validation.handleLfArchiveError)
       _ <- engine
         .validatePackages(packages.toMap)
-        .left
-        .map(e => new IllegalArgumentException(e.message))
-        .toTry
+        .loggingValidation(Validation.handleLfEnginePackageError)
     } yield dar
 
   override def uploadDarFile(request: UploadDarFileRequest): Future[UploadDarFileResponse] =
     withEnrichedLoggingContext(logging.submissionId(request.submissionId)) {
       implicit loggingContext =>
         logger.info("Uploading DAR file")
+        val submissionId = submissionIdGenerator(request.submissionId)
+        val darInputStream = new ZipInputStream(request.darFile.newInput())
 
         implicit val telemetryContext: TelemetryContext =
           DefaultTelemetry.contextFromGrpcThreadLocalContext()
 
-        val submissionId = submissionIdGenerator(request.submissionId)
-        val darInputStream = new ZipInputStream(request.darFile.newInput())
+        implicit val contextualizedErrorLogger: ContextualizedErrorLogger =
+          new DamlContextualizedErrorLogger(
+            logger,
+            loggingContext,
+            Some(submissionId),
+          )
 
         val response = for {
-          dar <- decodeAndValidate(darInputStream).fold(
-            err =>
-              Future.failed(
-                ValidationLogger
-                  .logFailureWithContext(
-                    request,
-                    errorFactories.invalidArgument(None)(err.getMessage)(
-                      new DamlContextualizedErrorLogger(
-                        logger,
-                        loggingContext,
-                        Some(submissionId),
-                      )
-                    ),
-                  )
-              ),
-            Future.successful,
-          )
+          dar <- Future.fromTry(decodeAndValidate(darInputStream))
           _ <- synchronousResponse.submitAndWait(submissionId, dar)
         } yield {
           for (archive <- dar.all) {
             logger.info(s"Package ${archive.getHash} successfully uploaded")
           }
           UploadDarFileResponse()
         }
-
         response.andThen(logger.logErrorsOnCall[UploadDarFileResponse])
     }
+
+  private implicit class ErrorValidations[E, R](result: Either[E, R]) {
+    def loggingValidation(validate: E => BaseError.Impl): Try[R] =
+      result.left.map { err =>
+        val baseError = validate(err)
+        baseError.log()
```

Comment:
Could this error be one that logs on creation and we could have double logging?

Reply:
`BaseError.Impl` is not logging on creation. Indeed someone could override this when defining the error implementation, but we don't have such implementations so far. The only errors that log on creation atm are the `LoggingTransactionErrorImpl`